openedx/core/djangoapps/content_tagging/rules.py

```diff
@@ -6,35 +6,169 @@
+def is_org_admin(orgs: list[Organization], user: UserType) -> bool:
     """
-    Returns True if the given user is a Taxonomy User for the given content taxonomy.
+    Return True if the given user is an admin for any of the given orgs.
+    """
+    return len(get_admin_orgs(user, orgs)) > 0
+
+
+def is_org_content_creator(orgs: list[Organization], user: UserType) -> bool:
+    """
+    Return True if the given user is a content creator for any of the given orgs.
+    """
+    return len(get_content_creator_orgs(user, orgs)) > 0
+
+
+def is_org_instructor(orgs: list[Organization], user: UserType) -> bool:
+    """
+    Return True if the given user is an instructor for any of the given orgs.
+    """
+    return len(get_instructor_orgs(user, orgs)) > 0
+
+
+def get_admin_orgs(user: UserType, orgs: list[Organization]) -> list[Organization]:
+    """
+    Returns a list of orgs that the given user is an admin, from the given list of orgs.
+    """
+    return [
+        org for org in orgs if OrgStaffRole(org=org.short_name).has_user(user)
+    ]
+
+
+def get_content_creator_orgs(user: UserType, orgs: list[Organization]) -> list[Organization]:
+    """
+    Returns a list of orgs that the given user is a content creator, the given list of orgs.
+    """
+    return [
+        org for org in orgs if (
+            OrgLibraryUserRole(org=org.short_name).has_user(user) or
+            OrgInstructorRole(org=org.short_name).has_user(user) or
+            OrgContentCreatorRole(org=org.short_name).has_user(user)
+        )
+    ]
+
+
+def get_instructor_orgs(user: UserType, orgs: list[Organization]) -> list[Organization]:
+    """
+    Returns a list of orgs that the given user is an instructor, from the given list of orgs.
+    """
+    instructor_roles = CourseAccessRole.objects.filter(
+        org__in=orgs,
+        user=user,
+        role__in=(CourseStaffRole.ROLE, CourseInstructorRole.ROLE),
+    )
+    return [role.org for role in instructor_roles]
+
+
+@rules.predicate
+def can_view_taxonomy(user: UserType, taxonomy: oel_tagging.Taxonomy) -> bool:
+    """
+    Returns True if the given user can view the given taxonomy.
+
+    Taxonomy admins can view any taxonomy.
+    Org-level staff can view any taxonomy that is associated with one of their orgs.
+    Org-level course creators and instructors can view any enabled taxonomy that is owned by one of their orgs.
+    """
+    # The following code allows METHOD permission (GET) in the viewset for everyone
+    if not taxonomy:
+        return True
+
+    taxonomy = taxonomy.cast()
+
+    # Taxonomy admins can view any taxonomy
+    if oel_tagging.is_taxonomy_admin(user):
+        return True
+
+    is_all_org = TaxonomyOrg.objects.filter(
+        taxonomy=taxonomy,
+        org=None,
+        rel_type=TaxonomyOrg.RelType.OWNER,
+    ).exists()
+
+    # Only taxonomy admins can view disabled all org taxonomies
+    if is_all_org and not taxonomy.enabled:
+        return False
```

Comment:
I'm not sure how we're supposed to handle a Taxonomy that's linked to "all orgs" _and_ linked to one or more specific orgs -- which takes precedence, the "all orgs" row, or the specific org rows?

I would guess that "all orgs" takes precedence.

So I'd expect this check to look like this, and `is_all_org` to not be referenced again in this function:

```suggestion
    # Only taxonomy admins can view disabled all org taxonomies
    if is_all_org:
        return taxonomy.enabled
```

Reply:
I agree that "all orgs" should take precedence.

We can't change the check that way because we would enable an ordinary user (without org) to see this taxonomy.  We can short circuit a `False` here (denying permission) but to return a `True` we need additional checks.

But reviewing the code, I think we don't need to check `is_all_org` after:

```python
# Org-level staff can view any taxonomy that is associated with one of their orgs.
if is_org_admin(taxonomy_orgs, user):
    return True
```

Reply:
@rpenido I don't think your conclusion ^ is correct?

> We can't change the check that way because we would enable an ordinary user (without org) to see this taxonomy. 

[This comment](https://github.com/openedx/modular-learning/issues/129#issuecomment-1747326670) says:

> * Enabled Taxonomy linked to all orgs - any registered user can view the taxonomy and its tags (3)
> * Disabled Taxonomy linked to all orgs - only global staff can view the taxonomy and its tags (5)

So my original suggestion stands:

```suggestion
    # Enabled all-org taxonomies can be viewed by any registered user.
    if is_all_org:
        return taxonomy.enabled
```

Reply:
You're right @pomegranited !
Done: dd4696a110c9d22baa1780fc8fcf6d63438cef01

I'll check/fix the related test tomorrow